src/Agent.Listener/Configuration/AutoLogonConfigurationManager.cs

```diff
@@ -0,0 +1,283 @@
+#if OS_WINDOWS
+using Microsoft.VisualStudio.Services.Agent.Util;
+using Microsoft.TeamFoundation.DistributedTask.WebApi;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Security.Principal;
+using System.Threading;
+
+namespace Microsoft.VisualStudio.Services.Agent.Listener.Configuration
+{
+    [ServiceLocator(Default = typeof(AutoLogonConfigurationManager))]
+    public interface IAutoLogonConfigurationManager : IAgentService
+    {
+        void Configure(CommandSettings command);
+        void UnConfigure();
+        bool RestartNeeded();
+        bool IsAutoLogonConfigured();
+    }
+
+    public class AutoLogonConfigurationManager : AgentService, IAutoLogonConfigurationManager
+    {
+        private ITerminal _terminal;
+
+        public override void Initialize(IHostContext hostContext)
+        {
+            base.Initialize(hostContext);
+            _terminal = hostContext.GetService<ITerminal>();
+        }
+
+        public void Configure(CommandSettings command)
+        {
+            AssertAdminAccess(false);
+
+            var logonAccount = command.GetAutoLogonUserName();
+
+            string domainName;
+            string userName;
+
+            GetAccountSegments(logonAccount, out domainName, out userName);
+
+            if ((string.IsNullOrEmpty(domainName) || domainName.Equals(".", StringComparison.CurrentCultureIgnoreCase)) && !logonAccount.Contains("@"))
+            {
+                logonAccount = String.Format("{0}\\{1}", Environment.MachineName, userName);
+            }
+
+            Trace.Info("LogonAccount after transforming: {0}, user: {1}, domain: {2}", logonAccount, userName, domainName);
+
+            string logonPassword = string.Empty;
+            var windowsServiceHelper = HostContext.GetService<INativeWindowsServiceHelper>();
+            while (true)
+            {
+                logonPassword = command.GetWindowsLogonPassword(logonAccount);
+                if (windowsServiceHelper.IsValidCredential(domainName, userName, logonPassword))
+                {
+                    Trace.Info("Credential validation succeeded");
+                    break;
+                }
+                
+                if (command.Unattended)
+                {
+                    throw new SecurityException(StringUtil.Loc("InvalidWindowsCredential"));
+                }
+                    
+                Trace.Info("Invalid credential entered");
+                _terminal.WriteLine(StringUtil.Loc("InvalidWindowsCredential"));
+            }
+
+            bool isCurrentUserSameAsAutoLogonUser = windowsServiceHelper.HasActiveSession(domainName, userName);                
+            var securityIdForTheUser = windowsServiceHelper.GetSecurityId(domainName, userName);
+            var regManager = HostContext.GetService<IWindowsRegistryManager>();
```

Comment:
why we have IWindowsRegistryManager and AutoLogonRegistryManager? should we merge them, since IWindowsRegistryManager is only used by autologonregmanager? the only reason you what them separated would be you can to unit test AutoLogonRegManger.